File: javascript_code/line/algo/sort/BubbleSorter.js


// 时间复杂度：O(n^2)
// 空间复杂度：O(1)
const sort = function (data) {
    if (data == null || data.length == 0) return

    for (let round = 1; round <= data.length; round++) { // 控制冒泡轮数
        let hasSwap = false
        const compareTimes = data.length - round
        for (let i = 0; i < compareTimes; i++) { // 控制每轮比较次数
            if (data[i] > data[i + 1]) {
                [data[i], data[i + 1]] = [data[i + 1], data[i]]
                hasSwap = true
            }
        }
        if (!hasSwap) break
    }
}

data = [12, 23, 36, 9, 24, 42]
sort(data)
console.log(data)

File: cpp_code/line/algo/sort/BubbleSorter.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include <algorithm>
#include <iostream>
#include <sstream>

using namespace std;

// 时间复杂度：O(n^2)
// 空间复杂度：O(1)
void sort(int* data, int length) {
    if (data == nullptr || length <= 1) return;

    for (int round = 1; round <= length; ++round) { // 控制冒泡轮数
        bool hasSwap = false;
        int compareTimes = length - round;
        for (int i = 0; i < compareTimes; ++i) { // 控制每轮比较次数
            if (data[i] > data[i + 1]) {
                swap(data[i], data[i + 1]);
                hasSwap = true;
            }
        }
        if (!hasSwap) break;
    }
}

int main() {
    int data[] = {12, 23, 36, 9, 24, 42};
    sort(data, 6);
    stringstream fmt;
    for (int i = 0; i < 6; ++i) {
        fmt << to_string(data[i]) << ",";
    }
    cout << fmt.str() << endl;
    return 0;
}